tests/objects/test_label_structure.py

```diff
@@ -0,0 +1,948 @@
+        height=0.2,
+        width=0.1,
+        top_left_x=0.5,
+        top_left_y=0.4,
+    )
+
+    object_instance_1.set_for_frame(coordinates=coordinates_1, frame=1)
+    object_instance_2.set_for_frame(coordinates=coordinates_2, frame=2)
+    object_instance_2.set_for_frame(coordinates=coordinates_2, frame=3)
+
+    label_row.add_object(object_instance_1)
+    label_row.add_object(object_instance_2)
+
+    objects = label_row.get_objects()
+    assert objects[0].object_hash == object_instance_1.object_hash
+    assert objects[1].object_hash == object_instance_2.object_hash
+    # The FE may sometimes rely on the order of these.
+
+    label_row.remove_object(object_instance_1)
+    objects = label_row.get_objects()
+    assert len(objects) == 1
+    assert objects[0].object_hash == object_instance_2.object_hash
+
+
+def test_filter_for_objects():
+    label_row = LabelRowClass(empty_image_group_labels, all_types_structure)
+    label_box = ObjectInstance(box_ontology_item)
+    label_polygon = ObjectInstance(polygon_ontology_item)
+
+    box_coordinates = BoundingBoxCoordinates(
+        height=0.1,
+        width=0.2,
+        top_left_x=0.3,
+        top_left_y=0.4,
+    )
+    polygon_coordinates = PolygonCoordinates(
+        values=[
+            PointCoordinate(x=0.2, y=0.1),
+            PointCoordinate(x=0.3, y=0.2),
+            PointCoordinate(x=0.5, y=0.3),
+            PointCoordinate(x=0.6, y=0.5),
+        ]
+    )
+
+    label_box.set_for_frame(box_coordinates, 1)
+    label_box.set_for_frame(box_coordinates, 2)
+    label_polygon.set_for_frame(polygon_coordinates, 2)
+    label_polygon.set_for_frame(polygon_coordinates, 3)
+
+    label_row.add_object(label_box)
+    label_row.add_object(label_polygon)
+
+    objects = label_row.get_objects()
+    assert len(objects) == 2
+
+    objects = label_row.get_objects(filter_ontology_object=polygon_ontology_item)
+    assert len(objects) == 1
+    assert objects[0].object_hash == label_polygon.object_hash
+
+    objects = label_row.get_objects(filter_ontology_object=box_ontology_item)
+    assert len(objects) == 1
+    assert objects[0].object_hash == label_box.object_hash
+
+    objects = label_row.get_objects(filter_ontology_object=polyline_ontology_item)
+    assert len(objects) == 0
+
+
+def test_add_wrong_coordinates():
+    label_box = ObjectInstance(box_ontology_item)
+    with pytest.raises(LabelRowError):
+        label_box.set_for_frame(POLYGON_COORDINATES, frame=1)
+
+
+def test_get_object_instances_by_frames():
+    label_row = LabelRowClass(empty_image_group_labels, all_types_structure)
+    label_box = ObjectInstance(box_ontology_item)
+    label_polygon = ObjectInstance(polygon_ontology_item)
+
+    label_box.set_for_frame(BOX_COORDINATES, 1)
+    label_box.set_for_frame(BOX_COORDINATES, 2)
+    label_polygon.set_for_frame(POLYGON_COORDINATES, 2)
+    label_polygon.set_for_frame(POLYGON_COORDINATES, 3)
+
+    label_row.add_object(label_box)
```

Comment:
We just had some issues with labels-with-objects-on-frames-past-end-of-video today; which makes me think: should we have the info about "available" frame numbers on the `label_row` here and then only allow the `set_for_frames` *after* the object is attached to a label row, so that we can check the validity?

We can obviously just throw from `add_object` but it will be less granular, potentially.

Reply:
Yes! this is actually in my latest change. You will always have the member of `number_of_frames`. Once you run `initialise_labels` (which might be named differently soon) you will be able to set stuff and then we will do the right checks already.